src/v/storage/disk_log_impl.cc

```diff
@@ -1731,6 +1734,244 @@ disk_log_impl::make_cached_reader(log_reader_config config) {
       .then([this](auto rdr) { return _readers_cache->put(std::move(rdr)); });
 }
 
+namespace details {
+// This accumulator is used to compute size of the on-disk representation of the
+// record batches
+struct batch_size_accumulator {
+    ss::future<ss::stop_iteration> operator()(model::record_batch b) {
+        vassert(
+          result_size_bytes != nullptr,
+          "batch_size_accumulator is not initialized properly");
+        // Target is exclusive:
+        // 'target' offset corresponds to the base offset of the
+        // batch that shouldn't be added to the result:
+        //
+        //            target---v
+        //     |++++++++++++++[X      ]           |
+        //
+        // Target is inclusive:
+        // 'target' offset corresponds to the last offset of the
+        // batch that should be added to the final result:
+        //
+        //                  target---v
+        //     |++++++++++++++[+++++++]X          |
+        //
+        if (target_is_inclusive) {
+            if (b.last_offset() > target) {
+                co_return ss::stop_iteration::yes;
+            }
+            *result_size_bytes += model::packed_record_batch_header_size
+                                  + b.data().size_bytes();
+            co_return ss::stop_iteration::no;
+        } else {
+            if (b.base_offset() >= target) {
+                co_return ss::stop_iteration::yes;
+            }
+            *result_size_bytes += model::packed_record_batch_header_size
+                                  + b.data().size_bytes();
+            co_return ss::stop_iteration::no;
+        }
+    }
+    bool end_of_stream() const { return false; }
+
+    size_t* result_size_bytes{nullptr};
+    model::offset target;
+    bool target_is_inclusive;
+};
+} // namespace details
+
+ss::future<size_t> disk_log_impl::get_file_offset(
+  ss::lw_shared_ptr<segment> s,
+  segment_index::entry index_entry,
+  model::offset target,
+  bool is_inclusive,
+  ss::io_priority_class priority) {
+    size_t size_bytes{index_entry.filepos};
+    details::batch_size_accumulator acc{
+      .result_size_bytes = &size_bytes,
+      .target = target,
+      .target_is_inclusive = is_inclusive,
+    };
+
+    storage::log_reader_config reader_cfg(index_entry.offset, target, priority);
+
+    reader_cfg.skip_batch_cache = true;
+    reader_cfg.skip_readers_cache = true;
+
+    auto reader = co_await make_reader(reader_cfg);
+
+    try {
+        co_await std::move(reader).consume(acc, model::no_timeout);
```

Comment:
i thought that consume() allowed the consumer to return a value, which would mean, you probably don't need to pass in a pointer to size_bytes in order to get size_bytes back out after the consumer is done processing data from the reader?